src/Middleware/HeaderPropagation/src/HeaderPropagationMessageHandler.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    /// <summary>
+    /// A message handler for propagating headers collected by the <see cref="HeaderPropagationMiddleware"/> to a outgoing request.
+    /// </summary>
+    public class HeaderPropagationMessageHandler : DelegatingHandler
+    {
+        private readonly HeaderPropagationValues _values;
+        private readonly HeaderPropagationOptions _options;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="HeaderPropagationMessageHandler"/>.
+        /// </summary>
+        /// <param name="options">The options that define which headers are propagated.</param>
+        /// <param name="values">The values of the headers to be propagated populated by the
+        /// <see cref="HeaderPropagationMiddleware"/>.</param>
+        public HeaderPropagationMessageHandler(IOptions<HeaderPropagationOptions> options, HeaderPropagationValues values)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _options = options.Value;
+
+            _values = values ?? throw new ArgumentNullException(nameof(values));
+        }
+
+        /// <summary>
+        /// Sends an HTTP request to the inner handler to send to the server as an asynchronous operation, after adding
+        /// the propagated headers.
+        /// </summary>
+        /// <remarks>
+        /// If an header with the same name is already present in the request, even if empty, the corresponding
+        /// propagated header will not be added.
+        /// </remarks>
+        /// <param name="request">The HTTP request message to send to the server.</param>
+        /// <param name="cancellationToken">A cancellation token to cancel operation.</param>
+        /// <returns>The task object representing the asynchronous operation.</returns>
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            foreach ((var headerName, var entry) in _options.Headers)
+            {
+                var outputName = string.IsNullOrEmpty(entry?.OutboundHeaderName) ? headerName : entry.OutboundHeaderName;
+
+                if (!request.Headers.Contains(outputName) &&
+                    _values.Headers.TryGetValue(headerName, out var values) &&
+                    !StringValues.IsNullOrEmpty(values))
+                {
+                    request.Headers.TryAddWithoutValidation(outputName, (string[])values);
```

Comment:
The cast `(string[])values` will cause an array allocation for most headers (which are only one)

Better would be to check the count and then switch on overload e.g.
```csharp
var count = values.Count;
if (count == 1)
{
    request.Headers.TryAddWithoutValidation(outputName, (string)values);
}
else if (count > 1)
{
    request.Headers.TryAddWithoutValidation(outputName, (string[])values);
}
```

Reply:
Also HttpClient has an odd behaviour where some headers are rejected from the Headers collection and instead need to be added to the Content.Headers collection (if there is a content); so might need to be more like:
```csharp
var hasContent = request.Content != null;

// ...

var count = values.Count;
if (count == 1)
{
    var value = (string)values;
    if (!request.Headers.TryAddWithoutValidation(outputName, value) && hasContent)
    {
        request.Content.Headers.TryAddWithoutValidation(outputName, value);
    }
} 
else ...
```
from: https://github.com/aspnet/Benchmarks/blob/7b1a5e986f06ae79270b13866e701cc6cb5bb395/src/Proxy/ProxyExtensions.cs#L31-L34

Reply:
Thanks @benaadams ! 
I'll raise a PR to address this by the end of the week.